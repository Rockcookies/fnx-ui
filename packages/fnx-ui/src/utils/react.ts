import { Children, ReactElement } from 'react';
import { isFragment } from 'react-is';

export const toElementArray = (
	children: React.ReactNode,
	option?: { keepEmpty?: boolean },
): ReactElement[] => {
	const keepEmpty = option && option.keepEmpty;
	let ret: React.ReactElement[] = [];

	Children.forEach(children, (child: any) => {
		if ((child === undefined || child === null) && !keepEmpty) {
			return;
		}

		if (Array.isArray(child)) {
			ret = ret.concat(toElementArray(child));
		} else if (isFragment(child) && child.props) {
			ret = ret.concat(toElementArray(child.props.children, option));
		} else {
			ret.push(child);
		}
	});

	return ret;
};
