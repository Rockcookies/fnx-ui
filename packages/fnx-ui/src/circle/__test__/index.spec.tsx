import { render } from '@testing-library/react';
import React from 'react';
import Circle from '../index';
import TestsDOM from '../../../test/dom';

describe('<Circle/>', () => {
	it('change with animation', () => {
		const { container, rerender } = render(<Circle progress={70}></Circle>);

		expect(container).toMatchSnapshot();

		rerender(<Circle progress={80}></Circle>);

		expect(container).toMatchSnapshot();
	});
	it('should change circle size when using size prop', () => {
		const { container } = render(
			<Circle size={300} progress={70}></Circle>,
		);

		expect(container).toMatchSnapshot();
	});

	it('Progress bar color, input object format can define gradient color', () => {
		const gradientColor = {
			'0%': '#3fecff',
			'100%': '#6149f6',
		};

		const { container } = render(
			<Circle
				className="Circle"
				progress={70}
				strokeColor={gradientColor}
			></Circle>,
		);

		expect(container).toMatchSnapshot();
	});
	it('should change stroke linecap when using stroke-linecap prop', () => {
		const { container } = render(
			<Circle strokeLinecap={'square'} progress={70}></Circle>,
		);

		expect(container).toMatchSnapshot();
	});

	it('should change gap position when using gapPosition prop', () => {
		const { container } = render(
			<>
				<Circle gapPosition="top" progress={70}></Circle>,
				<Circle gapPosition="bottom" progress={70}></Circle>,
				<Circle gapPosition="left" progress={70}></Circle>,
				<Circle gapPosition="right" progress={70}></Circle>,
			</>,
		);

		expect(container).toMatchSnapshot();
	});

	it('should change strokeColor between gradient and color string correctly', () => {
		const { container, rerender } = render(
			<Circle
				strokeColor={{
					'0%': '#3fecff',
					'100%': '#6149f6',
				}}
				progress={70}
			></Circle>,
		);

		const track = TestsDOM.mustQuerySelector(
			container,
			'.fnx-circle__hover',
		);

		expect((track as any).style.stroke).toBe('');

		rerender(<Circle strokeColor="red" progress={70}></Circle>);

		expect((track as any).style.stroke).toBe('red');

		rerender(
			<Circle
				strokeColor={{
					'0%': '#3fecff',
					'100%': '#6149f6',
				}}
				progress={70}
			></Circle>,
		);

		expect((track as any).style.stroke).toBe('');
	});
});
