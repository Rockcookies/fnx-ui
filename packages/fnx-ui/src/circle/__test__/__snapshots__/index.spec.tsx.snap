// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Circle/> Progress bar color, input object format can define gradient color 1`] = `
<div>
  <div
    class="fnx-circle Circle"
  >
    <svg
      viewBox="0 0 100 100"
    >
      <defs>
        <lineargradient
          id="fnx-circle-linearGradient-3"
          x1="100%"
          x2="0%"
          y1="0%"
          y2="0%"
        >
          <stop
            offset="0%"
            stop-color="#3fecff"
            stop-opacity="1"
          />
          <stop
            offset="100%"
            stop-color="#6149f6"
            stop-opacity="1"
          />
        </lineargradient>
      </defs>
      <path
        class="fnx-circle__trail"
        d="M 50,50 m 0,-47.5
   a 47.5,47.5 0 1 ,0 0,95
   a 47.5,47.5 0 1 ,0 0,-95"
        fill-opacity="0"
        stroke-linecap="round"
        stroke-width="5"
        style="stroke-dasharray: 298.45130209103036px 298.45130209103036px; stroke-dashoffset: -0px; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s;"
      />
      <path
        class="fnx-circle__hover"
        d="M 50,50 m 0,-47.5
   a 47.5,47.5 0 1 ,1 0,95
   a 47.5,47.5 0 1 ,1 0,-95"
        fill-opacity="0"
        opacity="1"
        stroke="url(#fnx-circle-linearGradient-3)"
        stroke-linecap="round"
        stroke-width="5"
        style="stroke-dasharray: 208.91591146372124px 298.45130209103036px; stroke-dashoffset: -0px; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; transition-duration: 0.3s, 0.3s, 0.3s, 0.06s;"
      />
    </svg>
    <div
      class="fnx-circle__text"
    />
  </div>
</div>
`;

exports[`<Circle/> change with animation 1`] = `
<div>
  <div
    class="fnx-circle"
  >
    <svg
      viewBox="0 0 100 100"
    >
      <path
        class="fnx-circle__trail"
        d="M 50,50 m 0,-47.5
   a 47.5,47.5 0 1 ,0 0,95
   a 47.5,47.5 0 1 ,0 0,-95"
        fill-opacity="0"
        stroke-linecap="round"
        stroke-width="5"
        style="stroke-dasharray: 298.45130209103036px 298.45130209103036px; stroke-dashoffset: -0px; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s;"
      />
      <path
        class="fnx-circle__hover"
        d="M 50,50 m 0,-47.5
   a 47.5,47.5 0 1 ,1 0,95
   a 47.5,47.5 0 1 ,1 0,-95"
        fill-opacity="0"
        opacity="1"
        stroke="#1989fa"
        stroke-linecap="round"
        stroke-width="5"
        style="stroke-dasharray: 208.91591146372124px 298.45130209103036px; stroke-dashoffset: -0px; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; transition-duration: 0.3s, 0.3s, 0.3s, 0.06s;"
      />
    </svg>
    <div
      class="fnx-circle__text"
    />
  </div>
</div>
`;

exports[`<Circle/> change with animation 2`] = `
<div>
  <div
    class="fnx-circle"
  >
    <svg
      viewBox="0 0 100 100"
    >
      <path
        class="fnx-circle__trail"
        d="M 50,50 m 0,-47.5
   a 47.5,47.5 0 1 ,0 0,95
   a 47.5,47.5 0 1 ,0 0,-95"
        fill-opacity="0"
        stroke-linecap="round"
        stroke-width="5"
        style="stroke-dasharray: 298.45130209103036px 298.45130209103036px; stroke-dashoffset: -0px; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s;"
      />
      <path
        class="fnx-circle__hover"
        d="M 50,50 m 0,-47.5
   a 47.5,47.5 0 1 ,1 0,95
   a 47.5,47.5 0 1 ,1 0,-95"
        fill-opacity="0"
        opacity="1"
        stroke="#1989fa"
        stroke-linecap="round"
        stroke-width="5"
        style="stroke-dasharray: 238.7610416728243px 298.45130209103036px; stroke-dashoffset: -0px; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; transition-duration: 0s, 0s;"
      />
    </svg>
    <div
      class="fnx-circle__text"
    />
  </div>
</div>
`;

exports[`<Circle/> should change circle size when using size prop 1`] = `
<div>
  <div
    class="fnx-circle"
    style="width: 300px; height: 300px;"
  >
    <svg
      viewBox="0 0 100 100"
    >
      <path
        class="fnx-circle__trail"
        d="M 50,50 m 0,-47.5
   a 47.5,47.5 0 1 ,0 0,95
   a 47.5,47.5 0 1 ,0 0,-95"
        fill-opacity="0"
        stroke-linecap="round"
        stroke-width="5"
        style="stroke-dasharray: 298.45130209103036px 298.45130209103036px; stroke-dashoffset: -0px; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s;"
      />
      <path
        class="fnx-circle__hover"
        d="M 50,50 m 0,-47.5
   a 47.5,47.5 0 1 ,1 0,95
   a 47.5,47.5 0 1 ,1 0,-95"
        fill-opacity="0"
        opacity="1"
        stroke="#1989fa"
        stroke-linecap="round"
        stroke-width="5"
        style="stroke-dasharray: 208.91591146372124px 298.45130209103036px; stroke-dashoffset: -0px; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; transition-duration: 0.3s, 0.3s, 0.3s, 0.06s;"
      />
    </svg>
    <div
      class="fnx-circle__text"
    />
  </div>
</div>
`;

exports[`<Circle/> should change gap position when using gapPosition prop 1`] = `
<div>
  <div
    class="fnx-circle"
  >
    <svg
      viewBox="0 0 100 100"
    >
      <path
        class="fnx-circle__trail"
        d="M 50,50 m 0,-47.5
   a 47.5,47.5 0 1 ,0 0,95
   a 47.5,47.5 0 1 ,0 0,-95"
        fill-opacity="0"
        stroke-linecap="round"
        stroke-width="5"
        style="stroke-dasharray: 298.45130209103036px 298.45130209103036px; stroke-dashoffset: -0px; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s;"
      />
      <path
        class="fnx-circle__hover"
        d="M 50,50 m 0,-47.5
   a 47.5,47.5 0 1 ,1 0,95
   a 47.5,47.5 0 1 ,1 0,-95"
        fill-opacity="0"
        opacity="1"
        stroke="#1989fa"
        stroke-linecap="round"
        stroke-width="5"
        style="stroke-dasharray: 208.91591146372124px 298.45130209103036px; stroke-dashoffset: -0px; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; transition-duration: 0.3s, 0.3s, 0.3s, 0.06s;"
      />
    </svg>
    <div
      class="fnx-circle__text"
    />
  </div>
  ,
  <div
    class="fnx-circle"
  >
    <svg
      viewBox="0 0 100 100"
    >
      <path
        class="fnx-circle__trail"
        d="M 50,50 m 0,47.5
   a 47.5,47.5 0 1 ,0 0,-95
   a 47.5,47.5 0 1 ,0 0,95"
        fill-opacity="0"
        stroke-linecap="round"
        stroke-width="5"
        style="stroke-dasharray: 298.45130209103036px 298.45130209103036px; stroke-dashoffset: -0px; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s;"
      />
      <path
        class="fnx-circle__hover"
        d="M 50,50 m 0,47.5
   a 47.5,47.5 0 1 ,1 0,-95
   a 47.5,47.5 0 1 ,1 0,95"
        fill-opacity="0"
        opacity="1"
        stroke="#1989fa"
        stroke-linecap="round"
        stroke-width="5"
        style="stroke-dasharray: 208.91591146372124px 298.45130209103036px; stroke-dashoffset: -0px; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; transition-duration: 0.3s, 0.3s, 0.3s, 0.06s;"
      />
    </svg>
    <div
      class="fnx-circle__text"
    />
  </div>
  ,
  <div
    class="fnx-circle"
  >
    <svg
      viewBox="0 0 100 100"
    >
      <path
        class="fnx-circle__trail"
        d="M 50,50 m -47.5,0
   a 47.5,47.5 0 1 ,0 95,0
   a 47.5,47.5 0 1 ,0 -95,0"
        fill-opacity="0"
        stroke-linecap="round"
        stroke-width="5"
        style="stroke-dasharray: 298.45130209103036px 298.45130209103036px; stroke-dashoffset: -0px; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s;"
      />
      <path
        class="fnx-circle__hover"
        d="M 50,50 m -47.5,0
   a 47.5,47.5 0 1 ,1 95,0
   a 47.5,47.5 0 1 ,1 -95,0"
        fill-opacity="0"
        opacity="1"
        stroke="#1989fa"
        stroke-linecap="round"
        stroke-width="5"
        style="stroke-dasharray: 208.91591146372124px 298.45130209103036px; stroke-dashoffset: -0px; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; transition-duration: 0.3s, 0.3s, 0.3s, 0.06s;"
      />
    </svg>
    <div
      class="fnx-circle__text"
    />
  </div>
  ,
  <div
    class="fnx-circle"
  >
    <svg
      viewBox="0 0 100 100"
    >
      <path
        class="fnx-circle__trail"
        d="M 50,50 m 47.5,0
   a 47.5,47.5 0 1 ,0 -95,0
   a 47.5,47.5 0 1 ,0 95,0"
        fill-opacity="0"
        stroke-linecap="round"
        stroke-width="5"
        style="stroke-dasharray: 298.45130209103036px 298.45130209103036px; stroke-dashoffset: -0px; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s;"
      />
      <path
        class="fnx-circle__hover"
        d="M 50,50 m 47.5,0
   a 47.5,47.5 0 1 ,1 -95,0
   a 47.5,47.5 0 1 ,1 95,0"
        fill-opacity="0"
        opacity="1"
        stroke="#1989fa"
        stroke-linecap="round"
        stroke-width="5"
        style="stroke-dasharray: 208.91591146372124px 298.45130209103036px; stroke-dashoffset: -0px; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; transition-duration: 0.3s, 0.3s, 0.3s, 0.06s;"
      />
    </svg>
    <div
      class="fnx-circle__text"
    />
  </div>
  ,
</div>
`;

exports[`<Circle/> should change stroke linecap when using stroke-linecap prop 1`] = `
<div>
  <div
    class="fnx-circle"
  >
    <svg
      viewBox="0 0 100 100"
    >
      <path
        class="fnx-circle__trail"
        d="M 50,50 m 0,-47.5
   a 47.5,47.5 0 1 ,0 0,95
   a 47.5,47.5 0 1 ,0 0,-95"
        fill-opacity="0"
        stroke-linecap="square"
        stroke-width="5"
        style="stroke-dasharray: 298.45130209103036px 298.45130209103036px; stroke-dashoffset: -0px; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s;"
      />
      <path
        class="fnx-circle__hover"
        d="M 50,50 m 0,-47.5
   a 47.5,47.5 0 1 ,1 0,95
   a 47.5,47.5 0 1 ,1 0,-95"
        fill-opacity="0"
        opacity="1"
        stroke="#1989fa"
        stroke-linecap="square"
        stroke-width="5"
        style="stroke-dasharray: 208.91591146372124px 298.45130209103036px; stroke-dashoffset: -0px; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; transition-duration: 0.3s, 0.3s, 0.3s, 0.06s;"
      />
    </svg>
    <div
      class="fnx-circle__text"
    />
  </div>
</div>
`;
